## Krypto Blatt 3
---

*Fynn Jonathan Rühlemann und Dennis Schielke*

## 13)

a)

k=16: 

$2^{k-1}$ = 32768

$2^k$ = 65536


k=24:

$2^{k-1}$ = 8388608

$2^k$ = 16777216

In [83]:
import math

# x1/ln(x1) - x2/ln(x2)

# k=16
expected16 = (2**16/math.log(2**16)) - (2**15/math.log(2**15))
print("Erwartete Primzahlen für k=16:", expected16)

# k=24
expected16 = (2**24/math.log(2**24)) - (2**23/math.log(2**23))
print("Erwartete Primzahlen für k=24:", expected16)

Erwartete Primzahlen für k=16: 2757.663480824557
Erwartete Primzahlen für k=24: 482334.18397527665


- Erwartete Primzahlen für k=16: 2757.663480824557
- Erwartete Primzahlen für k=24: 482334.18397527665

In [4]:
def getPrimes(end):
    primes = set()
    for x in range(1, end+1):
        if(checkPrime(x)):
            primes.add(x)
    return primes

def checkPrime(x):
    if(x == 2):
        return True
    if((x % 2 == 0) or (x <= 1)):
        return False
    for i in range(3, int(math.sqrt(x)+1), 2):
        if((x % i == 0)):
            return False
    return True


set161 = getPrimes(2**16)
set162 = getPrimes(2**15)

set241 = getPrimes(2**24)
set242 = getPrimes(2**23)

print("Menge an Primzahlen zwischen 2^16 und 2^16-1:", len(set161 - set162))
print("Menge an Primzahlen zwischen 2^24 und 2^24-1:", len(set241 - set242))


    

Menge an Primzahlen zwischen 2^16 und 2^16-1: 3030
Menge an Primzahlen zwischen 2^24 und 2^24-1: 513708


- Menge an Primzahlen zwischen 2^16 und 2^16-1: 3030
- Menge an Primzahlen zwischen 2^24 und 2^24-1: 513708

## 14)

In [28]:
def gcdt(zahl1, zahl2):
    for i in range(zahl2):
        if(zahl1 % (i+2) == 0 and zahl2 % (i+2) == 0):
            return False
    return True

In [30]:
import random

num1 = 2**80 + 85
num2 = 2**80 + 71

def fermat(N):
    for i in range(47):
        a = random.randint(2, N-2)

        if(math.gcd(a, N) != 1):
            print("N ist keine Primzahl")
            return
        
        b = a**(N-1) % N

        if(b != 1):
            print("N ist keine Primzahl")
            return
        
        print("N ist wahrscheinlich eine Primzahl")
        return
    
print("N = 2^80 + 85:", fermat(num1))
print("N = 2^80 + 71:", fermat(num2))

KeyboardInterrupt: 

## 15)

In [89]:
def SmallestMillerRabinWitness(N):
    # Miller-Rabin-Test
    if(N % 2 == 0):
        return False
    kAndM = getKAndM(N)
    k = kAndM[0]
    m = kAndM[1]

    a = random.randint(2, N-2)
    if(math.gcd(a, N) != 1):
        return False
    b = a**m % N
    i = 0
    if(b == 1):
        return True
    
    while(i <= k):
        b = b**2 % N
        i += 1

    if (b != 1):
        return False
    return True

def getKAndM(N):
    k = 0
    while(2**k < (N-1)):
        k += 1
    k -= 1

    def checkK(checkNum):
        m = 1
        while(checkNum*m <= (N-1)):
            if(checkNum*m == (N-1)):
                return m
            m += 1
        return -1
    
    while True:
        checkNum = 2**k
        m = checkK(checkNum)
        if(m != -1):
            break
        k -= 1

    return (k, m)

#### a) 

Zuerst müssen wir den kleinsten Miller-Rabin-Zeugen für N = 3215031751 finden um zu zeigen, dass N keine Primzahl ist. 
Dies können wir ganz leicht machen:

In [90]:
def durchlauf(N):
    for i in range(N-1):
        if(math.gcd(i+2, N) != 1):
            if(SmallestMillerRabinWitness(i+2)==False):
                print(i+2)
                return

durchlauf(3215031751)

302


Wie wir sehen können ist der kleinste Miller-Rabin-Zeuge für N die Zahl 302.

#### b)

In [97]:
def durchlauf2(N):
    for i in range(N-1):
        if(math.gcd(i+2, N) == 1):
            if(SmallestMillerRabinWitness(i+2)==False):
                print(i+2)

In [96]:
durchlauf2(145)

10
20
30
35
40
50
55
58
60
70
80
90
100
105
110
116
120
130
140


In [ ]:
durchlauf2(1105)

## Aufgabe 16
---

#### a) 

Zeigen Sie: falls N eine Carmichaelzahl ist, dann ist $p_i$ - 1 Teiler von N - 1 für alle $1 <= i <= 3$

$N = p_1 * p_2 * p_3$

Beispiel: 

$p_1 = 7, p_2 = 13, p_3 = 19$

$ 7 * 13 * 19 = 1729$

$ 1729 - 1 = 1728$

$ p_1 = 6, p_2 = 12, p_3 = 18 = 


In [40]:
print(1728%6)
print(1728%12)
print(1728%18)

0
0
0


Es ist zu sehen, dass diese Behauptung für ein gewähltes Beispiel stimmt. Da die Definition einer Carmichaelzahl darauf zutrifft, dass $p_1 - 1$ Teiler von N-1 sein muss für jedes $i$ gilt diese Behauptung für eine bel. Carmichaelzahl

#### b) 

Zeigen Sie: falls $p_i$ - 1 Teiler von N - 1 für alle $1 <= i <= 3$ ist, dann ist N eine Carmichaelzahl

Dies gilt wieder aus der Definition der Carmichaelzahl.

#### c)

Dies wird im folgenden durch den gegebenen Code gezeigt:

In [45]:
def carmichaelzahl(p1, p2, p3):
    N = p1*p2*p3
    if((N-1)%(p1-1) == 0 and (N-1)%(p2-1) == 0 and (N-1)%(p3-1) == 0):
        return True
    return False

def findCar():
    for i in range(1000):
        m1 = 6*(i+1)+1
        m2 = 12*(i+1)+1
        m3 = 18*(i+1)+1
        if(is_prime(m1) and is_prime(m2) and is_prime(m3)):
            print("Carmichaelzahl:", 6*(i+1)+1, 12*(i+1)+1, 18*(i+1)+1)
            if(carmichaelzahl(m1, m2, m3)==True):
                print("True")
           
def is_prime(n):
    for i in range(2,n):
        if (n%i) == 0:
            return False
    return True

findCar()

Carmichaelzahl: 7 13 19
True
Carmichaelzahl: 37 73 109
True
Carmichaelzahl: 211 421 631
True
Carmichaelzahl: 271 541 811
True
Carmichaelzahl: 307 613 919
True
Carmichaelzahl: 331 661 991
True
Carmichaelzahl: 337 673 1009
True
Carmichaelzahl: 601 1201 1801
True
Carmichaelzahl: 727 1453 2179
True
Carmichaelzahl: 1171 2341 3511
True
Carmichaelzahl: 1237 2473 3709
True
Carmichaelzahl: 1297 2593 3889
True
Carmichaelzahl: 1531 3061 4591
True
Carmichaelzahl: 1657 3313 4969
True
Carmichaelzahl: 2221 4441 6661
True
Carmichaelzahl: 2281 4561 6841
True
Carmichaelzahl: 2557 5113 7669
True
Carmichaelzahl: 3037 6073 9109
True
Carmichaelzahl: 3061 6121 9181
True
Carmichaelzahl: 3067 6133 9199
True
Carmichaelzahl: 4261 8521 12781
True
Carmichaelzahl: 4447 8893 13339
True
Carmichaelzahl: 4801 9601 14401
True
Carmichaelzahl: 4951 9901 14851
True
Carmichaelzahl: 5227 10453 15679
True
Carmichaelzahl: 5581 11161 16741
True
Carmichaelzahl: 5851 11701 17551
True


Dieser Code zeigt, dass dies für eine beliebige m worauf die Vorraussetzungen zutreffen gilt. 

#### d)

Durch den Einsetzen der kleinsten Zahlen findet man heraus, dass durch die Formel (6m+1)(12m+1)(18m+1) die Zahl 6 direkt eine Carmichaelzahl ist, da:

- (6*6+1) = 37 = Primzahl
- (12*6+1) = 73 = Primzahl
- (18*6+1) = 109 = Primzahl

Dadurch ist die 6 eine Carmichaelzahl